jina/peapods/compoundpod/__init__.py

```diff
@@ -0,0 +1,398 @@
+        return self.args.pod_role
+
+    @property
+    def is_singleton(self) -> bool:
+        """
+        Return if the Pod contains only a single Pea
+
+        :return: true if there is only one replica, else false
+        """
+        return not (self.is_head_router or self.is_tail_router)
+
+    @property
+    def name(self) -> str:
+        """
+        The name of this :class:`CompoundPod`.
+
+        :return: name of the pod
+        """
+        return self.args.name
+
+    @property
+    def port_expose(self) -> int:
+        """
+        Get the grpc port number.
+
+        :return: exposed port
+        """
+        return self.first_pea_args.port_expose
+
+    @property
+    def host(self) -> str:
+        """
+        Get the host name of this Pod.
+
+        :return: host name
+        """
+        return self.first_pea_args.host
+
+    @property
+    def host_in(self) -> str:
+        """
+        Get the host_in of this pod.
+
+        :return: host name of incoming requests
+        """
+        return self.head_args.host_in
+
+    @property
+    def host_out(self) -> str:
+        """
+        Get the host_out of this pod.
+
+        :return: host name of outgoing requests
+        """
+        return self.tail_args.host_out
+
+    @property
+    def address_in(self) -> str:
+        """
+        Get the full incoming address of this pod.
+
+        :return: address for incoming requests
+        """
+        return f'{self.head_args.host_in}:{self.head_args.port_in} ({self.head_args.socket_in!s})'
+
+    @property
+    def address_out(self) -> str:
+        """
+        Get the full outgoing address of this pod.
+
+        :return: address for outgoing requests.
+        """
+        return f'{self.tail_args.host_out}:{self.tail_args.port_out} ({self.tail_args.socket_out!s})'
+
+    @property
+    def first_pea_args(self) -> Namespace:
+        """
+        Return the first non-head/tail pea's args
+
+        :return: arguments of the first pea which is not a head pea or tail pea
+        """
+        # note this will be never out of boundary
+        return self.replicas_args['replicas'][0]
+
+    def _parse_args(
+        self, args: Namespace
+    ) -> Dict[str, Optional[Union[List[Namespace], Namespace]]]:
+        replicas_args = {'head': None, 'tail': None, 'replicas': []}
+        if getattr(args, 'replicas', 1) > 1:
+            # reasons to separate head and tail from peas is that they
+            # can be deducted based on the previous and next pods
+            _set_after_to_pass(args)
+            self.is_head_router = True
+            self.is_tail_router = True
+            replicas_args['head'] = _copy_to_head_args(args, PollingType.ANY.is_push)
+            replicas_args['tail'] = _copy_to_tail_args(args)
+            replicas_args['replicas'] = _set_pod_args(
+                args, replicas_args['head'], replicas_args['tail']
+            )
+
+        # TODO this case does properly not exist anymore
```

Comment:
Should this line be removed? What's this about?

Reply:
nice catch - removed